src/stream/src/task/barrier_manager/managed_state.rs

```diff
@@ -40,6 +42,70 @@ use crate::executor::monitor::StreamingMetrics;
 use crate::executor::Barrier;
 use crate::task::ActorId;
 
+fn pending_for_once() -> impl Future<Output = ()> + Unpin + 'static {
+    let mut polled = false;
+    poll_fn(move |_cx| {
+        if polled {
+            Poll::Ready(())
+        } else {
+            polled = true;
+            Poll::Pending
+        }
+    })
+}
+
+fn instrument_barrier(
+    barrier_await_tree_reg: &Option<Arc<Mutex<await_tree::Registry<u64>>>>,
+    barrier: &Barrier,
+) -> Option<InflightBarrierInstrumented> {
+    barrier_await_tree_reg.as_ref().map(|r| {
+        let mut ret = InflightBarrierInstrumented(
+            r.lock()
+                .register(
+                    barrier.epoch.prev,
+                    format!(
+                        "Inflight Barrier prev = {}, curr = {}",
+                        barrier.epoch.prev, barrier.epoch.curr
+                    ),
+                )
+                .instrument(async {
+                    pending_for_once()
+                        .instrument_await("AwaitBarrierCollected")
+                        .await;
+                    pending_for_once().instrument_await("AwaitSync").await;
+                })
```

Comment:
Wow, this looks too hack to me.

TBH await-tree is designed to reflect the execution of futures in a natural and intuitive way. If we need to put in a lot of effort to mimic the spans using dummy futures because of how the event-loop executes, I would recommend discontinuing this functionality since it goes against the design goal. 🥲

Reply:
I think await tree is an ideal tool to trace the progress of any ongoing tasks. The key idea of this PR is to leverage its functionality to trace inflight barrier without involving too much extra works. Given that await tree doesn't expose lower layer control on the TreeContext and the inner poll state, the current way is the best I can do. 

I think we can have the current PR merged if we don't want to lose the useful ability to track ongoing barriers. In the future if await-tree supports lower layer control, such as `tree_root.enter("CollectBarrier")`, `tree_root.exit()`, we can then change to use the lower layer control api.

Reply:
I have removed this fake future and changed to monitor only the async sync_epoch method. @BugenZhao PTAL